### PR TITLE
Docker

### PR DESCRIPTION
**Pull request by sychan**
Dockerize UJS using jetty as the container

### CODE REVIEW

File: deployment/conf/.templates/deployment.cfg.templ [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+server-threads={{ default .Env.server_threads "20" }}
+min-memory={{ default .Env.min_memory "1000" }}
+max-memory={{ default .Env.max_memory "2000" }}
+workspace-url={{ default .Env.workspace_url "https://ci.kbase.us/services/ws" }}
+
+# Document server name. Used for logging.
+doc-server-name = {{ default .Env.doc_server_name "UJSDocServ" }}
+doc-server-docs-location = {{ default .Env.doc_server_docs_location "/server_docs" }}
+
+# Credentials of the account the workspace server should use to get user
```
Comment (MrCreosote) [inline deployment/conf/.templates/deployment.cfg.templ]: This seems out of place.

File: src/us/kbase/userandjobstate/UserAndJobStateServer.java [hunk 2]
```diff
@@ -146,7 +146,7 @@
 	private static final String GIT =
 			"https://github.com/kbase/user_and_job_state";
 	
-	private static final String VER = "0.2.2";
+	private static final String VER = "0.2.3";
```
Comment (MrCreosote) [inline src/us/kbase/userandjobstate/UserAndJobStateServer.java:149]: Need to update in the tests as well.

**Review by MrCreosote — CHANGES_REQUESTED** [review]

File: Dockerfile [hunk 3]
```diff
@@ -0,0 +1,31 @@
+FROM kbase/kb_jre
+
+# These ARGs values are passed in via the docker build command
+ARG BUILD_DATE
+ARG VCS_REF
+ARG BRANCH=develop
+
+COPY deployment/ /kb/deployment/
```
Comment (scanon) [inline Dockerfile:25]: Can we make this a multistage build before we merge.

**Review by scanon — CHANGES_REQUESTED** [review]

File: Dockerfile (hunk 3 above)
Comment (sychan) [reply Dockerfile:25]: Its been converted to multistage dockerhub builds

File: src/us/kbase/userandjobstate/UserAndJobStateServer.java (hunk 2 above)
Comment (sychan) [reply src/us/kbase/userandjobstate/UserAndJobStateServer.java:149]: Fixed in JSONRPCLayerTest.java:129

**Comment by sychan** [comment]
@MrCreosote I believe that all issues have been resolved, please recheck.

File: Dockerfile [hunk 4]
```diff
@@ -0,0 +1,48 @@
+FROM kbase/sdkbase2:latest as build
```
Comment (MrCreosote) [inline Dockerfile:1]: Is this just for the sdk compile step? It doesn't look like there's an sdk compile going on here, so AFAICT it's unnecessary.

Or is it just for pod2html? If that's the case I'd just skip that and just copy the other relevant commands from build-docs here.

File: deployment/conf/.templates/deployment.cfg.templ [hunk 5]
```diff
@@ -0,0 +1,23 @@
+[UserAndJobState]
+mongodb-database={{ default .Env.mongodb_database "userjobstate" }}
+mongodb-pwd={{ default .Env.mongodb_pwd "" }}
+mongodb-host={{ default .Env.mongodb_host "ci-mongo" }}
+mongodb-user={{ default .Env.mongodb_user "ujsserv" }}
+service-port={{ default .Env.service_port "8080" }}
+basedir={{ default .Env.basedir "userandjobstate" }}
```
Comment (MrCreosote) [inline deployment/conf/.templates/deployment.cfg.templ]: There is no basedir param for ujs

File: deployment/conf/.templates/deployment.cfg.templ [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+kbase-admin-token={{ default .Env.kbase_admin_token "" }}
+server-threads={{ default .Env.server_threads "20" }}
+min-memory={{ default .Env.min_memory "1000" }}
+max-memory={{ default .Env.max_memory "2000" }}
```
Comment (MrCreosote) [inline deployment/conf/.templates/deployment.cfg.templ]: I think server-threads and memory params only work if you're using the glassfish init script IIRC

File: Dockerfile [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+      maintainer="Steve Chan sychan@lbl.gov"
+
+EXPOSE 7058
+ENTRYPOINT [ "/kb/deployment/bin/dockerize" ]
+CMD [ "-template", "/kb/deployment/conf/.templates/deployment.cfg.templ:/kb/deployment/conf/deployment.cfg", \
+      "-template", "/kb/deployment/conf/.templates/http.ini.templ:/kb/deployment/jettybase/start.d/http.ini", \
+      "-template", "/kb/deployment/conf/.templates/server.ini.templ:/kb/deployment/jettybase/start.d/server.ini", \
+      "-template", "/kb/deployment/conf/.templates/start_server.sh.templ:/kb/deployment/bin/start_server.sh", \
+      "-stdout", "/kb/deployment/jettybase/logs/request.log", \
+      "/kb/deployment/bin/start_server.sh" ]
```
Comment (MrCreosote) [inline Dockerfile:45]: Are we still using a start server script? All my services just have a jetty directive right in the CMD.

File: deployment/conf/.templates/server.ini.templ [hunk 8]
```diff
@@ -0,0 +1,62 @@
+# --------------------------------------- 
+# Module: server
+--module=server
+
+### ThreadPool configuration
+## Minimum number of threads
+jetty.threadPool.minThreads={{ default .Env.jetty_minThreads "5" }}
+
+## Maximum number of threads
+# jetty.threadPool.maxThreads={{ default .Env.jetty_maxThreads "20" }}
```
Comment (MrCreosote) [inline deployment/conf/.templates/server.ini.templ:10]: Just fyi, unlike WS, there's no reason to not use the default jetty thread count (which is like 500 or something)

Comment (MrCreosote) [reply deployment/conf/.templates/server.ini.templ:10]: Also note this line is commented out. Is that intentional?

File: deployment/conf/.templates/start_server.sh.templ [hunk 9]
```diff
@@ -0,0 +1,10 @@
+#!/bin/sh
+#
+
+JAVA_OPTS="-Djava.awt.headless=true -server \
+               -Xms{{ default .Env.min_memory "1000" }}m -Xmx{{ default .Env.max_memory "3000" }}m \
+               -XX:+UseG1GC"
+
+# This script assumes that the docker base image includes a runnable JETTY environment
+# that provides $JETTY_HOME. Currently the base image is built on library/jetty:jre8
+java -DSTOP.PORT=8079 -DSTOP.KEY=foo -Djetty.home=$JETTY_HOME -jar $JETTY_HOME/start.jar
```
Comment (MrCreosote) [inline deployment/conf/.templates/start_server.sh.templ]: No need for the STOP* stuff

File: RELEASE_NOTES.txt [hunk 10]
```diff
@@ -5,6 +5,13 @@ OVERVIEW
 A service for 1) maintaining global KBase user state, and 2) maintaining
 job state. See the API docs for more details.
 
+VERSION: 0.2.3 (2/13/18)
+-------------------------
+
+UPDATED FEATURES / MAJOR BUG FIXES:
+- Added makefile target to create docker image using jetty as servlet container
```
Comment (MrCreosote) [inline RELEASE_NOTES.txt]: There's no changes for the makefile?

**Review by MrCreosote — CHANGES_REQUESTED** [review]

File: Dockerfile (hunk 4 above)
Comment (sychan) [reply Dockerfile:1]: The "make build-libs" triggers an ant compile. We don't want the source repo or a full build java toolchain laying around in the runtime image, so I use the ant and jar commands in the sdk image to build the warfile.

File: deployment/conf/.templates/deployment.cfg.templ (hunk 5 above)
Comment (sychan) [reply deployment/conf/.templates/deployment.cfg.templ]: Deleted.

File: deployment/conf/.templates/deployment.cfg.templ (hunk 6 above)
Comment (sychan) [reply deployment/conf/.templates/deployment.cfg.templ]: I looked at the Jetty server.ini config, and the threads controlled by jetty_minThreads and jetty_maxThreads and the min_memory and max_memory are now applied in the start_server.sh template. I've removed these entries.

File: Dockerfile (hunk 7 above)
Comment (sychan) [reply Dockerfile:45]: The start server script provides a way to configure the memory the JVM uses, which we still want to control. It looks we could just set the _JAVA_OPTIONS environment variable, but its a little cryptic. I also feel the shell wrapper is a little more explicit and it also allows use to use shell variables like $JETTY_HOME in the command line

File: deployment/conf/.templates/server.ini.templ (hunk 8 above)
Comment (sychan) [reply deployment/conf/.templates/server.ini.templ:10]: Honestly, I can't remember if it was intentional or not anymore. If it doesn't effect memory or performance, we can just comment both of them out and use the defaults.

File: RELEASE_NOTES.txt (hunk 10 above)
Comment (sychan) [reply RELEASE_NOTES.txt]: The old PR used a travis build, the new dockerhub build didn't need the Makefile change. I've updated the release note accordingly.

**Comment by sychan** [comment]
@MrCreosote Please give it another look. I've deleted the issues that were outdated cruft, but several of the comments don't/shouldn't need to be changed.

File: Dockerfile (hunk 4 above)
Comment (MrCreosote) [reply Dockerfile:1]: Those aren't in kb_jre?

**Review by MrCreosote — COMMENTED** [review]
Just one question why the sdk image is needed for the 1st build stage vs. a jre image.

File: Dockerfile (hunk 4 above)
Comment (sychan) [reply Dockerfile:1]: No, its a kb_jre not kb_jdk

Comment (MrCreosote) [reply Dockerfile:1]: Seems weird we have to bring in all the SDK stuff just to get ant and java, but *shrug*

**Pull request merged**